airbyte-cdk/python/airbyte_cdk/sources/declarative/concurrent_declarative_source.py

```diff
@@ -0,0 +1,337 @@
+            # because their behavior can deviate from ConcurrentBehavior
+
+            slice_boundary_fields = (
+                declarative_cursor.get_partition_field_start().eval(config=config),
+                declarative_cursor.get_partition_field_end().eval(config=config),
+            )
+
+            interpolated_start_date = declarative_cursor.get_start_datetime()
+            start_date = interpolated_start_date.get_datetime(config=config)
+
+            interpolated_end_date = declarative_cursor.get_end_datetime()
+            end_date_provider = partial(interpolated_end_date.get_datetime, config) if interpolated_end_date else None
+
+            # DatetimeBasedCursor returns an isodate.Duration if step uses month or year precision. This still works in our
+            # code, but mypy may complain when we actually implement this in the concurrent low-code source. To fix this, we
+            # may need to convert a Duration to timedelta by multiplying month by 30 (but could lose precision).
+            step_length = declarative_cursor.get_step()
+
+            # The low-code DatetimeBasedCursor component uses the default max timedelta value which can lead to an
+            # OverflowError when building datetime intervals. We should proactively cap this to the current moment instead
+            if isinstance(step_length, datetime.timedelta) and step_length >= datetime.timedelta.max:
+                step_length = datetime.datetime.now(tz=datetime.timezone.utc) - start_date
+
+            return DeclarativeCursorAttributes(
+                cursor_field=CursorField(declarative_cursor.cursor_field.eval(config=config)),  # type: ignore # cursor_field is always cast to an interpolated string
+                datetime_format=declarative_cursor.datetime_format,
+                slice_boundary_fields=slice_boundary_fields,
+                start=start_date,  # type: ignore  # Having issues w/ inspection for GapType and CursorValueType as shown in existing tests. Confirmed functionality is working in practice
+                end_provider=end_date_provider,  # type: ignore  # Having issues w/ inspection for GapType and CursorValueType as shown in existing tests. Confirmed functionality is working in practice
+                slice_range=step_length,
+                lookback_window=parse_duration(declarative_cursor.lookback_window) if declarative_cursor.lookback_window else None,
+                cursor_granularity=parse_duration(declarative_cursor.cursor_granularity) if declarative_cursor.cursor_granularity else None,
+            )
+        return None
+
+    @staticmethod
+    def _stream_supports_concurrent_partition_processing(declarative_stream: DeclarativeStream, cursor: DatetimeBasedCursor) -> bool:
+        """
+        Many connectors make use of stream_state during interpolation on a per-partition basis under the assumption that
```

Comment:
Do we have documentation somewhere on things we need to do to fully migrate to the one CDK? If not, I can start one and this will be part of the list.

Reply:
We do not have it documented at a high level as far as I am aware. However, in my writeup for the breaking change in `cdk-migrations.md`, I've listed all the connectors in the repo that use `stream_state` in a non-thread safe way and what we can do to fix them.

But at a higher level, we should have a line item for the unification of our CDKs. Basically we should update our schemas to not use `stream_state` as an interpolation context and remove it wherever it's used from the Retriever and below (although I don't think it's actually used above that)

Reply:
I'll take some time to improve previous work on documenting this [here](https://whimsical.com/untangling-rfr-concurrent-low-code-VbjEeGwy72UeDEQDSWiMQ3). Thanks for sharing this!